### PR TITLE
refactor(aria.js): first steps

### PR DESCRIPTION
**Pull request by pkra**
Cf. w3c/aria-common#104 

Much work to be done.

* [x] f52f2bf - add a very simple test
  * `test.sh` runs respec on ARIA before/after applying aria.js changes and diffs the two results
* rough goals 
  * clarify (e.g., switch to const/let, replace reused variable names with unique ones)
  * simplify (e.g., loops over querySelectorAll, use sets instead of filtering)
  * revise HTML string creation
  * extract loops
* steps
  * [x] pdef, sdef loop
  * [x] generating HTML for indices of (global) states&properties as well as role=roletype
  * [x] rdef loop
  * [x] roleInfo loop
  * [x] propList loop
  * [x] generate HTML for role-related indices, sub-Roles
  * [x] clean-up
* continuation
  * [x] rewrite subRoles code
* TODO
  * [ ] review questions in buildInheritedStatesProperties
  * [ ] review TODOs
  * [ ] return roleInfo to make nodejs-generation easier.
* Future 
  * turn everything into ESM? [to modularize code]
  * review other JS files (to prepare their refactorings)

### CODE REVIEW

**Comment by pkra** [comment]
Another "fun" find: pdef/sdef and rdef are written to unnecessarily different HTML - p/sdef have an extra nested span inside the h4.

**Comment by pkra** [comment]
> Another "fun" find: pdef/sdef and rdef are written to unnecessarily different HTML - p/sdef have an extra nested span inside the h4.

I guess fixing #105 would add another complication but it feels still the right thing to create similar looking HTML in the long run.

**Comment by pkra** [comment]
Huh.

https://github.com/w3c/aria-common/blob/2fe4fe20cffad6314bddb6d6b9ed09c63c087cfb/script/aria.js#L216

and then for each rdef we push and item but also add a property with an array.

https://github.com/w3c/aria-common/blob/2fe4fe20cffad6314bddb6d6b9ed09c63c087cfb/script/aria.js#L280-L285

Sounds like we want an object here.

**Comment by pkra** [comment]
> Sounds like we want an object here.

But of course this once again breaks things in mysterious ways.

**Comment by pkra** [comment]
We'll need to change eslint config to set ecmaVersion to latest (so we can use optional chaining).

**Comment by pkra** [comment]
rdef loop refactoring seems complete (for now).

**Comment by pkra** [comment]
The  roleInfo loop is somewhat complete. The overloaded array pattern needs resolving eventually.

**Comment by pkra** [comment]
Ugh. The last change (meant to fix roleInfo output) has regressions in the HTML output.

**Comment by pkra** [comment]
I found a slightly hacky way to generate roleInfo.js using a "light" dom implementation in NodeJS, i.e., just using the raw index.html and aria.js. It could be made non-hacky if we decide to switch to ESM - but that would require small changes in the spec to allow for asynchronous code.

**Comment by pkra** [comment]
One less overloaded array construction.

**Comment by pkra** [comment]
@jnurthen, @spectranaut, I think this is as good as it gets for now. 

There are plenty of TODOs items noted throughout. I especially failed to really clarify (what is now called) `buildInheritedStatesProperties()`.  As noted in comments, I think I've identified a few bugs (both specifically to `buildInheritedStatesProperties` and more generally).

If you want me to change anything to make it easier to review, I'd be happy to.

**Comment by spectranaut** [comment]
Hey @pkra, I'm starting to look through all this code -- I never looked at the previous script, so this is really my first introduction, as I said :)

I appreciate your test script - it looks like there is no change in the resulting html before and after this refactor, when I run it!

I'll take a second closer look soon, but for this PR, do you have any specific goals for the review? Are you curious to just make sure it's readable, the javascript looks sound enough, at this stage? Is your goal to land this pretty close to as-is, so that the follow up work happens in sequential PRs?

**Comment by pkra** [comment]
> do you have any specific goals for the review?

I'd say my primary goals is to establish a fresh baseline we can work on together.

* a sanity check - does anything jump out as particularly problematic? Anything that could obviously be improved?
* is this approach ok? Do we want a completely different style?
* reach a point where we (i.e., multiple people) can discuss the remaining TODOs and future ideas
  * in particular those that would change the HTML and/or roleInfo.js 

As I said on the last call, I didn't feel knowledgeable enough to go the route of rewriting aria.js from scratch (in James words: because "we know what we want it to do" - even now I wouldn't say the same for myself).  

I think re-writing it again is a good idea in the long run - I have a lot of stupid ideas for that now - but should probably involve change the spec markup more heavily, e.g., James' idea of moving closer to "standard" respec markup.

Long story short, I'd be fine with landing this as is and doing more work later (which would give the benefit of a full rollout). But I'm also fine continuing work here.

File: script/aria.js [hunk 1]
```diff
@@ -7,746 +20,395 @@
...
+
+    const applicabilityText = container.querySelector(
+        "." + itemEntry.is + "-applicability"
+    ).innerText;
+    const isDefault = applicabilityText === "All elements of the base markup";
+    const isProhibited =
+        applicabilityText ===
+        "All elements of the base markup except for some roles or elements that prohibit its use";
+    const isDeprecated =
+        applicabilityText === "Use as a global deprecated in ARIA 1.2";
```
Comment (spectranaut) [inline script/aria.js:64]: I see, this is awful. Definitely, it makes more sense to start with a JSON object rather than looking for this text..

File: script/aria.js [hunk 2]
```diff
@@ -7,746 +20,395 @@
...
-            }
+/**
+ *
+ * @param {HTMLElement} container - parent of sdef or pdef or rdef
+ */
+const rewriteDefContainer = (container) => {
+    // if we are in a div, convert that div to a section
+    // TODO:
+    // a) seems to be always the case.
+    // b) Why don't we author the spec this way?
```
Comment (spectranaut) [inline script/aria.js:84]: Seems like we should. Anything that is not generated should just be in the static html. It's odd because this container has `id=<attributename>` on it for the url fragment, which probably should be generated, since it is the same as the pdef/sdef

File: script/aria.js [hunk 3]
```diff
@@ -7,746 +20,395 @@
...
+ * Replaces sdef/pdef with desired HTML
+ * @param {Object} propList -
+ * @param {HTMLElement} item - sdef or pdef, from nodeList.forEach
+ */
+const generateHTMLStatesAndProperties = function (propList, item) {
+    const title = item.getAttribute("title") || item.innerHTML;
+    const itemEntry = propList[title];
+    const dRef = item.nextElementSibling;
+    dRef.id = "desc-" + title; // TODO: too much of a side-effect?
+    dRef.setAttribute("role", "definition"); // TODO: ditto?
```
Comment (spectranaut) [inline script/aria.js:127]: this kind of stuff is a little weird to be adding here, it's not generated, it should just be in the html, it seems.

File: script/aria.js [hunk 4]
```diff
@@ -7,746 +20,395 @@
...
-                l = document.createElement("ul");
-                l.innerHTML = globalSPIndex;
-                node.replaceChild(l, parentNode);
-            }
-        }
+/**
+ *
+ * @param {HTMLElement} item - rdef element
+ */
+const rewriteRdef = function (item) {
```
Comment (spectranaut) [inline script/aria.js:102]: at least if this is functionally similar to `generateHTMLStatesAndProperties`, it should have a similar name?

File: script/aria.js [hunk 5]
```diff
@@ -764,16 +425,229 @@ function ariaAttributeReferences() {
...
+    // process the document before anything else is done
+    // first get the properties
+    const pdefsAndsdefs = document.querySelectorAll("pdef, sdef");
+    const pdefsAndsdefsContainer = [...pdefsAndsdefs].map(
+        (node) => node.parentNode
+    );
+
+    pdefsAndsdefs.forEach(populatePropList.bind(null, propList));
+    pdefsAndsdefs.forEach(populateGlobalSP.bind(null, propList, globalSP));
+    pdefsAndsdefs.forEach(generateHTMLStatesAndProperties.bind(null, propList));
```
Comment (spectranaut) [inline script/aria.js:600]: Each of these new function names could maybe follow a bit better of conventions... my top preference is at at least to rename any function that changes or produces html to have the prefix `render...`, and to describe specifically what it is creating -- like, `renderStatesAndPropertiesHeadings` instead of `generateHTMLStatesAndProperties`.

Also, related, I guess the "populate" is fine, but I might use a word like `build...` for anything that is just producing a data structure. Like `buildPropList`. 

Also.... I know you are reusing variable names, but personally I would use no short hands -- so, "buildStateAndPropertyList" instead of "buildPropList". and "buildGlobalStatesAndPropertiesList". I guess "pdefs" and "sdefs" etc are ok, considering they are using in our spec.

bu right now I still don't really understand why there is a `globalSP` and `propList` and the difference between these two things.. could they somehow be names a bit more clearly? Or are you trying to reuse these original names to make it easier to compare to the original script?

**Review by spectranaut — CHANGES_REQUESTED** [review]
Ok Peter I took another crack at going over this -- I wonder how much you would be interested in renaming functions or variables in this PR? :)

**Comment by pkra** [comment]
Thanks so much, @spectranaut! I'll need a little bit of time to get back into this.

**Comment by pkra** [comment]
Replaced by https://github.com/w3c/aria/pull/2408
